devicemanager testing: dynamically choose tmp dir

pohly opened the pull request:
This avoids the test issue #59488 that I was running into.

I believe I have a reasonable explanation for the race condition in that issue (TLDR: it's probably part of the gRPC API and k8s can only avoid the issue until a proper solution gets worked out together with gRPC), therefore I suggest to merge this PR now both because it avoids the issue and because using fixed tmp directories is something that should be avoided anyway.

/assign @jiayingz

pohly committed:
devicemanager testing: dynamically choose tmp dir

Hard-coding the tests to use /tmp/device_plugin for sockets is
problematic because it prevents running tests in parallel on the same
machine (perhaps because there are multiple developers, perhaps
because testing is done independently on different code checkouts).
/tmp/device_plugin also was not removed after testing.

This is probably not that relevant. But more importantly, this change
also fixes https://github.com/kubernetes/kubernetes/issues/59488.
"make test" failed in TestDevicePluginReRegistration because something
removed /tmp/device_plugin/device-plugin.sock while something else
tried to connect to it:

2018/02/07 14:34:39 Starting to serve on /tmp/device_plugin/device-plugin.sock
[pid 29568] connect(14, {sa_family=AF_UNIX, sun_path="/tmp/device_plugin/server.sock"}, 33) = 0
[pid 29568] unlinkat(AT_FDCWD, "/tmp/device_plugin/server.sock", 0) = 0
[pid 29568] unlinkat(AT_FDCWD, "/tmp/device_plugin/device-plugin.sock", 0) = 0
[pid 29568] --- SIGPIPE {si_signo=SIGPIPE, si_code=SI_USER, si_pid=29568, si_uid=1000} ---
[pid 29568] connect(6, {sa_family=AF_UNIX, sun_path="/tmp/device_plugin/device-plugin.sock"}, 40) = -1 ENOENT (No such file or directory)
E0207 14:34:39.961321   29568 endpoint.go:117] listAndWatch ended unexpectedly for device plugin mock with error rpc error: code = Unavailable desc = transport is closing
strace: Process 29623 attached
[pid 29574] connect(3, {sa_family=AF_UNIX, sun_path="/tmp/device_plugin/device-plugin.sock"}, 40) = -1 ENOENT (No such file or directory)
[pid 29623] connect(3, {sa_family=AF_UNIX, sun_path="/tmp/device_plugin/device-plugin.sock"}, 40) = -1 ENOENT (No such file or directory)
[pid 29574] connect(3, {sa_family=AF_UNIX, sun_path="/tmp/device_plugin/device-plugin.sock"}, 40) = -1 ENOENT (No such file or directory)
E0207 14:34:49.961324   29568 endpoint.go:60] Can't create new endpoint with path /tmp/device_plugin/device-plugin.sock err failed to dial device plugin: context deadline exceeded
E0207 14:34:49.961390   29568 manager.go:340] Failed to dial device plugin with request &RegisterRequest{Version:v1alpha2,Endpoint:device-plugin.sock,ResourceName:fake-domain/resource,}: failed to dial device plugin: context deadline exceeded
panic: test timed out after 2m0s

It's not entirely certain which code was to blame for this unlinkat()
calls (perhaps some cleanup code from a previous test running in a
goroutine?) but this no longer happened after switching to per-test
socket directories.

vikaschoudhary16 commented on pkg/kubelet/cm/devicemanager/manager_test.go:
```diff
@@ -40,25 +40,42 @@ import (
 )
 
+func tmpSocketDir() (socketDir, socketName, pluginSocketName string, err error) {
+	socketDir, err = ioutil.TempDir("", "device_plugin")
```
Would it be better to get a unique directory here using timestamp based suffix.
```golang
tmpDir, err := ioutil.TempDir("", "device_plugin"+fmt.Sprintf("%d", time.Now().Unix()))
```
And then use this in the setup(). Since there can be several setup() invocations in parallel and setup() calls `m.Start()` which eventually invokes removeContents, setup() should pass unique socket directory to the NewManagerImpl().

RenaudWasTaken replied:
**Tempdir:**
TempDir creates a new temporary directory in the directory dir with a name beginning with prefix and returns the path of the new directory. If dir is the empty string, TempDir uses the default directory for temporary files (see os.TempDir). Multiple programs calling TempDir simultaneously will not choose the same directory. It is the caller's responsibility to remove the directory when no longer needed.

vikaschoudhary16 reviewed (changes requested):
Thanks @pohly for the investigation. Your analysis seems correct to me.
About the fix in this PR, i have some suggestions to make. Please find inline.

pohly commented:
On Fri, 2018-02-09 at 01:02 -0800, Vikas Choudhary (vikasc) wrote:
> Would it be better to get a unique directory here using timestamp
> based suffix.
> tmpDir, err := ioutil.TempDir("", "device_plugin"+fmt.Sprintf("%d",
> time.Now().Unix()))
> And then use this in the setup(). Since there can be several setup()
> invocations in parallel and setup() calls m.Start() which eventually
> invokes removeContents, setup() should pass unique socket directory
> to the NewManagerImpl().

Are you concerned that multiple parallel calls to ioutil.TempDir() will
end up using the same directory? ioutil.TempDir() is explicitly
guaranteeing that this won't happen.

vikaschoudhary16 commented:
Ah yes, you are right. Thanks!

vikaschoudhary16 commented:
/ok-to-test

vikaschoudhary16 reviewed (commented):
/lgtm

vikaschoudhary16 commented:
/release-note-none
/area hw-accelerators

vikaschoudhary16 commented on pkg/kubelet/cm/devicemanager/manager_test.go:
```diff
@@ -86,8 +103,8 @@ func TestDevicePluginReRegistration(t *testing.T) {
 
-	p2 := NewDevicePluginStub(devs, pluginSocketName+".new")
-	err := p2.Start()
+	p2 := NewDevicePluginStub(devs, pluginSocketName+".TestDevicePluginReRegistration.new")
```
I dont think `TestDevicePluginReRegistration` is really needed.

pohly replied:
Right, this was an experimental change before changing the dir. Removed.

vikaschoudhary16 commented on pkg/kubelet/cm/devicemanager/manager_test.go:
```diff
@@ -114,7 +131,7 @@ func TestDevicePluginReRegistration(t *testing.T) {
 }
 
-func setup(t *testing.T, devs []*pluginapi.Device, callback monitorCallback) (Manager, *Stub) {
+func setup(t *testing.T, devs []*pluginapi.Device, callback monitorCallback, socketName string, pluginSocketName string) (Manager, *Stub) {
```
same as above

pohly replied:
Is this comment perhaps attached to the wrong line? If it was related to ".TestDevicePluginReRegistration.third", then I fixed that.

If the comment really was about the setup() line, then I don't understand what I should change here.

vikaschoudhary16 replied:
you got it right. comment on wrong line.

vikaschoudhary16 commented on pkg/kubelet/cm/devicemanager/manager_test.go:
```diff
@@ -86,8 +103,8 @@ func TestDevicePluginReRegistration(t *testing.T) {
 	devices := m.Devices()
 	require.Equal(t, 2, len(devices[testResourceName]), "Devices are not updated.")
```
Hope you wont mind adding couple of another minor changes as well. Can you please replace all `<-callbackChan` with : 
```golang
	select {
 	case <-callbackChan:
 		break
 	case <-time.After(time.Second):
 		t.FailNow()
 }
```
```

vikaschoudhary16 commented:
/lgtm cancel

pohly committed:
devicemanager testing: time out sooner

Each individual step should not take longer than a second.
Suggest by Vikas Choudhary (https://github.com/kubernetes/kubernetes/pull/59489#discussion_r167205672).

pohly commented:
@vikaschoudhary16 I think I have addressed your feedback in the force-pushed branch. Please have another look.

I also fixed a formatting issue (space instead of tab indention).

vikaschoudhary16 reviewed (changes requested):
>I0209 13:19:56.494] pkg/kubelet/cm/devicemanager/manager_test.go:105:9: undefined: time
I0209 13:19:56.495] pkg/kubelet/cm/devicemanager/manager_test.go:120:9: undefined: time
I0209 13:19:56.495] pkg/kubelet/cm/devicemanager/manager_test.go:137:9: undefined: time

You missed to import time.

RenaudWasTaken commented on pkg/kubelet/cm/devicemanager/manager_test.go:
```diff
@@ -40,25 +40,42 @@ import (
+
 func TestNewManagerImpl(t *testing.T) {
-	_, err := newManagerImpl(socketName)
+	socketDir, socketName, _, err := tmpSocketDir()
```
Wondering if we should have a [TestMain](http://cs-guy.com/blog/2015/01/test-main/) function here.
That would prevent writing this setup and teardown code each time.

What do you think?

pohly replied:
The problem in issue #59488 seemed to be that different tests in the same test program share the same socket dir and somehow interfere with each other. I am all ears if someone can explain why that happens, because it has not become clear to me.

If we use TestMain, we end up in the same situation and issue #59488 will probably occur again.

If it wasn't for that I'd agree with you, TestMain would be more suitable than the per-test boiler plate code.

pohly commented:
On Fri, 2018-02-09 at 16:00 +0000, Vikas Choudhary (vikasc) wrote:
> @vikaschoudhary16 requested changes on this pull request.
> I0209 13:19:56.494]
> pkg/kubelet/cm/devicemanager/manager_test.go:105:9: undefined: time
> I0209 13:19:56.495]
> pkg/kubelet/cm/devicemanager/manager_test.go:120:9: undefined: time
> I0209 13:19:56.495]
> pkg/kubelet/cm/devicemanager/manager_test.go:137:9: undefined: time
> You missed to import time.

Darn. Sorry for the noise. I'm working on a slower machine right now
and obviously took a shortcut :-/ It's fixed now.

vikaschoudhary16 commented:
/test pull-kubernetes-bazel-test

vikaschoudhary16 commented:
/lgtm

pohly commented:
/assign @jiayingz

RenaudWasTaken commented:
/lgtm

vikaschoudhary16 commented:
ping @ConnorDoyle @jiayingz .
Seems this should be merged.

jiayingz commented:
/lgtm
/approve

pohly commented:
/test pull-kubernetes-unit

k8s.io/kubernetes/test/integration/deployment TestRollbackDeploymentRSNoRevision failed. Looks like a flake. I'll file an issue once it succeeds.
